### PR TITLE
:sparkles: Add integration tests for both controllers

### LINKED ISSUE #73: Add integration tests for CSPO controllers
/kind feature

**Describe the solution you'd like**
We propose the introduction of integration tests for CSPO controllers.

**Acceptance Criteria:**
- [ ]  Add integration tests in the Go code
- [ ] The test setup need external dependencies.
- [ ]  Have Make target to run the tests easily.
- [ ]  Run the tests in the CI with GitHub Action.

### PR DESCRIPTION
**Pull request by michal-gubricky**
**What this PR does / why we need it**:
Adds integration tests for both CSPO controllers

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #73

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [X] add integration tests

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -55,6 +55,19 @@ jobs:
...
+          GIT_ORG_NAME: chess-knight
+          GIT_REPOSITORY_NAME: cluster-stacks
+          GO111MODULE: "on"
+          GIT_ACCESS_TOKEN: ${{ secrets.GIT_ACCESS_TOKEN }}
+          ENCODED_CLOUDS_YAML: ${{ secrets.ENCODED_CLOUDS_YAML }}
+        run: make test-integration
+
       # - name: Create Report
       #   run: make report-cover-html report-cover-treemap
       # - name: Running integration tests
```
Comment (chess-knight) [inline .github/workflows/test.yaml:73]: You can remove this comment probably

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -55,6 +55,19 @@ jobs:
...
 
+      - name: Running integration tests
+        env:
+          GIT_PROVIDER: github
+          # TODO: Change GIT_ORG_NAME to SovereignCloudStack
+          # when repo SovereignCloudStack/cluster-stack will
+          # have correct and stable releases for OpenStack
+          GIT_ORG_NAME: chess-knight
+          GIT_REPOSITORY_NAME: cluster-stacks
+          GO111MODULE: "on"
```
Comment (chess-knight) [inline .github/workflows/test.yaml:66]: Do we need this ?

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -55,6 +55,19 @@ jobs:
       - name: Running unit tests
         run: make test-unit
 
+      - name: Running integration tests
+        env:
+          GIT_PROVIDER: github
+          # TODO: Change GIT_ORG_NAME to SovereignCloudStack
+          # when repo SovereignCloudStack/cluster-stack will
+          # have correct and stable releases for OpenStack
+          GIT_ORG_NAME: chess-knight
```
Comment (chess-knight) [inline .github/workflows/test.yaml:64]: We can use `SovereignCloudStack` here probably

File: internal/test/helpers/envtest.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+	// DefaultPodNamespace is default the namespace for the envtest resources.
+	DefaultPodNamespace = "cspo-system"
+
+	// defaultKindClusterNodeImage is default node image for kind cluster.
+	defaultKindClusterNodeImage = "kindest/node:v1.26.6@sha256:6e2d8b28a5b601defe327b98bd1c2d1930b49e5d8c512e1895099e4504007adb" //#nosec
+)
+
+func init() {
+	// Calculate the scheme.
+	utilruntime.Must(clusterv1.AddToScheme(scheme))
```
Comment (chess-knight) [inline internal/test/helpers/envtest.go:86]: Why do we need a clusterv1 scheme?

**Comment by janiskemper** [comment]
I will review ones the review of @chess-knight is addressed! Just ping me again okay?

File: internal/test/helpers/envtest.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+	crdPaths := []string{
+		filepath.Join(root, "config", "crd", "bases"),
+	}
+
+	// Create the test environment.
+	env = &envtest.Environment{
+		Scheme:                scheme,
+		ErrorIfCRDPathMissing: true,
+		CRDDirectoryPaths:     crdPaths,
+		CRDs:                  []*apiextensionsv1.CustomResourceDefinition{},
```
Comment (chess-knight) [inline internal/test/helpers/envtest.go:103]: Can this be omitted?

File: .github/workflows/test.yaml (hunk 3 above)
Comment (michal-gubricky) [reply .github/workflows/test.yaml:64]: Yes, after you added the `node-images.yaml`  into [this release](https://github.com/SovereignCloudStack/cluster-stacks/releases/tag/openstack-scs-1-27-v2) yesterday. We can change `GIT_ORG_NAME` to `SovereignCloudStack`. I will also slightly adjust integration test, where we test that.

File: internal/test/integration/openstack/integration_controller_test.go
[the remark this answers is not in the document]
Comment (michal-gubricky) [reply internal/test/integration/openstack/integration_controller_test.go:1]: Yes, we can. But based on this, we should also rename a file `integration_test.go` inside the `integration/github` folder, right?

[the remark this answers is not in the document]
Comment (chess-knight) [reply internal/test/integration/openstack/integration_controller_test.go:1]: If I check CSO repository, I see [integration_test.go](https://github.com/SovereignCloudStack/cluster-stack-operator/tree/main/internal/test/integration/github) so IMO github can stay as it is

[the remark this answers is not in the document]
Comment (michal-gubricky) [reply internal/test/integration/openstack/integration_controller_test.go:1]: Yes, I also saw that. Based on the previous comment, both files are inside an integration folder and could be changed. But it is just a naming of files, which is not important in this case I think. I will change only the name of that one file as you suggested, to be consistent with the CSO repository.

[the remark this answers is not in the document]
Comment (chess-knight) [reply internal/test/integration/openstack/integration_controller_test.go:1]: Thanks, that is exactly what I meant

File: Makefile [hunk 6]
```diff
@@ -538,6 +542,18 @@ test-unit-openstack: $(SETUP_ENVTEST) $(GOTESTSUM) $(HELM)
 	CREATE_KIND_CLUSTER=false KUBEBUILDER_ASSETS="$(KUBEBUILDER_ASSETS)" $(GOTESTSUM) --junitfile=.coverage/junit.xml --format testname -- -mod=vendor \
 	-covermode=atomic -coverprofile=.coverage/cover.out -p=4 ./internal/controller/...
 
+.PHONY: test-integration-github
+test-integration-github: $(SETUP_ENVTEST) $(GOTESTSUM)
+	@mkdir -p $(shell pwd)/.coverage
+	CREATE_KIND_CLUSTER=false KUBEBUILDER_ASSETS="$(KUBEBUILDER_ASSETS)" $(GOTESTSUM) --junitfile=.coverage/junit.xml --format testname -- -mod=vendor \
```
[the remark this answers is not in the document]
Comment (michal-gubricky) [reply Makefile:548]: We do not need this for now and I will remove it. And also from the make target for `test-unit-openstack` where this was introduced first.

[the remark this answers is not in the document]
Comment (chess-knight) [reply Makefile:548]: Thanks, I think that CSO uses it for workload cluster integration tests, which we do not have

**Comment by michal-gubricky** [comment]
> I will review ones the review of @chess-knight is addressed! Just ping me again okay?

@janiskemper you can take a look at it now, all comments from @chess-knight  were addressed.

**Review by janiskemper — APPROVED** [review]
looks good!

**Review by chess-knight — APPROVED** [review]

**Pull request merged**